File: ios/Classes/Tasks/CalculateDistanceTask.m
//
//  CalculateDistanceTask.m
//  geolocator
//
//  Created by Maurits van Beusekom on 20/05/2019.
//

#import "CalculateDistanceTask.h"


@implementation CalculateDistanceTask {
    CLLocation *_startLocation;
    CLLocation *_endLocation;
}

- (instancetype)initWithContext:(TaskContext *)context completionHandler:(CompletionHandler)completionHandler {
    self = [super initWithContext:context completionHandler:completionHandler];
    if (self) {
        //
        [self parseCoordinates:context.arguments];
    }
    
    return self;
}

- (void)parseCoordinates:(id)arguments {
    if ([arguments isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees startLatitude = ((NSNumber *)arguments[@"startLatitude"]).doubleValue;
        CLLocationDegrees startLongitude = ((NSNumber *)arguments[@"startLongitude"]).doubleValue;
        CLLocationDegrees endLatitude = ((NSNumber *)arguments[@"endLatitude"]).doubleValue;
        CLLocationDegrees endLongitude = ((NSNumber *)arguments[@"endLongitude"]).doubleValue;
        
        _startLocation = [[CLLocation alloc] initWithLatitude:startLatitude longitude:startLongitude];
        _endLocation = [[CLLocation alloc] initWithLatitude:endLatitude longitude:endLongitude];
    } else {
        _startLocation = nil;
        _endLocation = nil;
    }
}

- (void)startTask {
    if (_startLocation == nil || _endLocation == nil) {
        [self handleErrorCode:@"ERROR_CALCULATE_DISTANCE_INVALID_PARAMS" message:@"Please supply start and end coordinates."];
        [self stopTask];
        return;
    }
    
    CLLocationDistance distance = [_endLocation distanceFromLocation:_startLocation];
    
    [[self context] resultHandler](@(distance));
    [self stopTask];
}

@end

File: ios/Classes/Tasks/CalculateDistanceTask.swift
//
//  CalculateDistance.swift
//  geolocator
//
//  Created by Maurits van Beusekom on 18/07/2018.
//

import Foundation
import CoreLocation

class CalculateDistanceTask : Task, TaskProtocol {
    private var _startLocation: CLLocation?
    private var _endLocation: CLLocation?
    
    required init(context: TaskContext, completionHandler: CompletionHandler?) {
        super.init(context: context,
                   completionHandler: completionHandler)
        
        parseCoordinates(arguments: context.arguments)
    }
    
    func parseCoordinates(arguments: Any?) {
        if let location = context.arguments as! NSDictionary? {
            let startLatitude = location.object(forKey: "startLatitude") as! CLLocationDegrees
            let startLongitude = location.object(forKey: "startLongitude") as! CLLocationDegrees
            let endLatitude = location.object(forKey: "endLatitude") as! CLLocationDegrees
            let endLongitude = location.object(forKey: "endLongitude") as! CLLocationDegrees
            
            _startLocation = CLLocation.init(latitude: startLatitude, longitude: startLongitude)
            _endLocation = CLLocation.init(latitude: endLatitude, longitude: endLongitude)
        } else {
            _startLocation = nil
            _endLocation = nil
        }
    }
    
    func startTask() {
        guard let startLocation = _startLocation, let endLocation = _endLocation else {
            handleError(
                code: "ERROR_CALCULATE_DISTANCE_INVALID_PARAMS",
                message: "Please supply start and end coordinates.")
            
            stopTask()
            return;
        }
        
        let distance = endLocation.distance(from: startLocation)
        
        context.resultHandler(distance)
        
        stopTask()
    }
}
